File: android-kit-base/src/test/java/com/mparticle/kits/KitIntegrationTest.java
package com.mparticle.kits;


import android.content.Context;
import android.util.SparseBooleanArray;

import com.mparticle.MParticle;
import com.mparticle.internal.MPUtility;
import com.mparticle.mock.MockMParticle;

import org.junit.Assert;
import org.junit.Test;
import org.mockito.Mockito;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

public class KitIntegrationTest {

    @Test
    public void testGetAllUserAttributesWithoutLists() throws Exception {
        MParticle.setInstance(new MockMParticle());
        KitIntegration integration = new KitIntegration() {
            @Override
            public String getName() {
                return null;
            }

            @Override
            protected List<ReportingMessage> onKitCreate(Map<String, String> settings, Context context) {
                return null;
            }

            @Override
            public List<ReportingMessage> setOptOut(boolean optedOut) {
                return null;
            }
        };
        integration.setConfiguration(Mockito.mock(KitConfiguration.class));

        Map<String, Object> attributes = new HashMap<>();
        attributes.put("key 1", "value 1");
        List<String> list = new LinkedList<>();
        list.add("value 2");
        list.add("value 3");
        attributes.put("key 2", list);

        Mockito.when(MParticle.getInstance().Identity().getCurrentUser().getUserAttributes()).thenReturn(attributes);
        Map<String, Object> filteredAttributes = integration.getAllUserAttributes();
        Assert.assertEquals("value 1", filteredAttributes.get("key 1"));
        Assert.assertEquals("value 2,value 3", filteredAttributes.get("key 2"));
    }

    class MockSparseBooleanArray extends SparseBooleanArray {
        @Override
        public boolean get(int key) {
            return get(key, false);
        }

        @Override
        public boolean get(int key, boolean valueIfKeyNotFound) {
            System.out.print("SparseArray getting: " + key);
            if (map.containsKey(key)) {
                return map.get(key);
            }else{
                return valueIfKeyNotFound;
            }
        }

        Map<Integer, Boolean> map = new HashMap<Integer, Boolean>();
        @Override
        public void put(int key, boolean value) {
            map.put(key, value);
        }

        @Override
        public void clear() {
            map.clear();
        }

        @Override
        public int size() {
            return map.size();
        }

        @Override
        public String toString() {
            return map.toString();
        }
    }

    @Test
    public void testGetAllUserAttributesWithoutListsWithFilters() throws Exception {
        MParticle.setInstance(new MockMParticle());
        KitIntegration integration = new KitIntegration() {
            @Override
            public String getName() {
                return null;
            }

            @Override
            protected List<ReportingMessage> onKitCreate(Map<String, String> settings, Context context) {
                return null;
            }

            @Override
            public List<ReportingMessage> setOptOut(boolean optedOut) {
                return null;
            }
        };
        KitConfiguration configuration = Mockito.mock(KitConfiguration.class);
        MockSparseBooleanArray mockArray = new MockSparseBooleanArray();
        mockArray.put(MPUtility.mpHash("key 4"), false);
        mockArray.put(MPUtility.mpHash("key 3"), false);
        Mockito.when(configuration.getUserAttributeFilters()).thenReturn(mockArray);
        integration.setConfiguration(configuration);

        Map<String, Object> attributes = new HashMap<>();
        attributes.put("key 1", "value 1");
        attributes.put("key 4", "value 4");
        List<String> list = new LinkedList<>();
        list.add("value 2");
        list.add("value 3");
        attributes.put("key 2", list);
        attributes.put("key 3", list);
        Mockito.when(MParticle.getInstance().Identity().getCurrentUser().getUserAttributes()).thenReturn(attributes);
        Map<String, Object> filteredAttributes = integration.getAllUserAttributes();
        Assert.assertEquals("value 1", filteredAttributes.get("key 1"));
        Assert.assertEquals("value 2,value 3", filteredAttributes.get("key 2"));
        Assert.assertNull(filteredAttributes.get("key 3"));
        Assert.assertNull(filteredAttributes.get("key 4"));
    }



    @Test
    public void testGetAllUserAttributesWithLists() throws Exception {
        MParticle.setInstance(new MockMParticle());
        KitIntegration integration = new AttributeListenerIntegration();
        integration.setConfiguration(Mockito.mock(KitConfiguration.class));

        Map<String, Object> attributes = new HashMap<>();
        attributes.put("key 1", "value 1");
        List<String> list = new LinkedList<>();
        list.add("value 2");
        list.add("value 3");
        attributes.put("key 2", list);

        Mockito.when(MParticle.getInstance().Identity().getCurrentUser().getUserAttributes()).thenReturn(attributes);
        Map<String, Object> filteredAttributes = integration.getAllUserAttributes();
        Assert.assertEquals("value 1", filteredAttributes.get("key 1"));
        Assert.assertEquals(list, filteredAttributes.get("key 2"));
    }

    @Test
    public void testGetAllUserAttributesWithListsAndFilters() throws Exception {
        MParticle.setInstance(new MockMParticle());
        KitIntegration integration = new AttributeListenerIntegration();
        KitConfiguration configuration = Mockito.mock(KitConfiguration.class);
        MockSparseBooleanArray mockArray = new MockSparseBooleanArray();
        mockArray.put(MPUtility.mpHash("key 4"), false);
        mockArray.put(MPUtility.mpHash("key 3"), false);
        Mockito.when(configuration.getUserAttributeFilters()).thenReturn(mockArray);
        integration.setConfiguration(configuration);

        Map<String, Object> attributes = new HashMap<>();
        attributes.put("key 1", "value 1");
        attributes.put("key 4", "value 4");
        List<String> list = new LinkedList<>();
        list.add("value 2");
        list.add("value 3");
        attributes.put("key 2", list);
        attributes.put("key 3", list);
        Mockito.when(MParticle.getInstance().Identity().getCurrentUser().getUserAttributes()).thenReturn(attributes);
        Map<String, Object> filteredAttributes = integration.getAllUserAttributes();
        Assert.assertEquals("value 1", filteredAttributes.get("key 1"));
        Assert.assertEquals(list, filteredAttributes.get("key 2"));
        Assert.assertNull(filteredAttributes.get("key 3"));
        Assert.assertNull(filteredAttributes.get("key 4"));
    }

    private class AttributeListenerIntegration extends KitIntegration implements KitIntegration.AttributeListener {

        @Override
        public void setUserAttribute(String attributeKey, String attributeValue) {

        }

        @Override
        public void setUserAttributeList(String attributeKey, List<String> attributeValueList) {

        }

        @Override
        public boolean supportsAttributeLists() {
            return true;
        }

        @Override
        public void setAllUserAttributes(Map<String, String> userAttributes, Map<String, List<String>> userAttributeLists) {

        }

        @Override
        public void removeUserAttribute(String key) {

        }

        @Override
        public void setUserIdentity(MParticle.IdentityType identityType, String identity) {

        }

        @Override
        public void removeUserIdentity(MParticle.IdentityType identityType) {

        }

        @Override
        public List<ReportingMessage> logout() {
            return null;
        }

        @Override
        public String getName() {
            return null;
        }

        @Override
        protected List<ReportingMessage> onKitCreate(Map<String, String> settings, Context context) {
            return null;
        }

        @Override
        public List<ReportingMessage> setOptOut(boolean optedOut) {
            return null;
        }
    }
}

File: android-core/src/test/kotlin/com/mparticle/kits/KitIntegrationTest.kt
package com.mparticle.kits

import android.content.Context
import android.util.SparseBooleanArray
import com.mparticle.MParticle
import com.mparticle.MParticle.IdentityType
import com.mparticle.internal.MPUtility
import com.mparticle.kits.KitIntegration.AttributeListener
import com.mparticle.mock.MockMParticle
import org.junit.Assert
import org.junit.Test
import org.mockito.Mockito
import java.util.LinkedList

class KitIntegrationTest {
    @Test
    @Throws(Exception::class)
    fun testGetAllUserAttributesWithoutLists() {
        MParticle.setInstance(MockMParticle())
        val integration: KitIntegration = object : KitIntegration() {
            override fun getName(): String? {
                return null
            }

            override fun onKitCreate(
                settings: Map<String, String>,
                context: Context
            ): List<ReportingMessage>? {
                return null
            }

            override fun setOptOut(optedOut: Boolean): List<ReportingMessage>? {
                return null
            }
        }
        integration.configuration = Mockito.mock(KitConfiguration::class.java)
        val attributes: MutableMap<String, Any> = HashMap()
        attributes["key 1"] = "value 1"
        val list: MutableList<String> = LinkedList()
        list.add("value 2")
        list.add("value 3")
        attributes["key 2"] = list
        Mockito.`when`(
            MParticle.getInstance()!!.Identity().currentUser!!.userAttributes
        ).thenReturn(attributes)
        val filteredAttributes = integration.allUserAttributes
        Assert.assertEquals("value 1", filteredAttributes["key 1"])
        Assert.assertEquals("value 2,value 3", filteredAttributes["key 2"])
    }

    internal inner class MockSparseBooleanArray : SparseBooleanArray() {
        override fun get(key: Int): Boolean {
            return get(key, false)
        }

        override fun get(key: Int, valueIfKeyNotFound: Boolean): Boolean {
            print("SparseArray getting: $key")
            return if (map.containsKey(key)) {
                map[key]!!
            } else {
                valueIfKeyNotFound
            }
        }

        var map: MutableMap<Int, Boolean> = HashMap()
        override fun put(key: Int, value: Boolean) {
            map[key] = value
        }

        override fun clear() {
            map.clear()
        }

        override fun size(): Int {
            return map.size
        }

        override fun toString(): String {
            return map.toString()
        }
    }

    @Test
    @Throws(Exception::class)
    fun testGetAllUserAttributesWithoutListsWithFilters() {
        MParticle.setInstance(MockMParticle())
        val integration: KitIntegration = object : KitIntegration() {
            override fun getName(): String? {
                return null
            }

            override fun onKitCreate(
                settings: Map<String, String>,
                context: Context
            ): List<ReportingMessage>? {
                return null
            }

            override fun setOptOut(optedOut: Boolean): List<ReportingMessage>? {
                return null
            }
        }
        val configuration = Mockito.mock(
            KitConfiguration::class.java
        )
        val mockArray = MockSparseBooleanArray()
        mockArray.put(MPUtility.mpHash("key 4"), false)
        mockArray.put(MPUtility.mpHash("key 3"), false)
        Mockito.`when`(configuration.userAttributeFilters).thenReturn(mockArray)
        integration.configuration = configuration
        val attributes: MutableMap<String, Any> = HashMap()
        attributes["key 1"] = "value 1"
        attributes["key 4"] = "value 4"
        val list: MutableList<String> = LinkedList()
        list.add("value 2")
        list.add("value 3")
        attributes["key 2"] = list
        attributes["key 3"] = list
        Mockito.`when`(
            MParticle.getInstance()!!.Identity().currentUser!!.userAttributes
        ).thenReturn(attributes)
        val filteredAttributes = integration.allUserAttributes
        Assert.assertEquals("value 1", filteredAttributes["key 1"])
        Assert.assertEquals("value 2,value 3", filteredAttributes["key 2"])
        Assert.assertNull(filteredAttributes["key 3"])
        Assert.assertNull(filteredAttributes["key 4"])
    }

    @Test
    @Throws(Exception::class)
    fun testGetAllUserAttributesWithLists() {
        MParticle.setInstance(MockMParticle())
        val integration: KitIntegration = AttributeListenerIntegration()
        integration.configuration = Mockito.mock(KitConfiguration::class.java)
        val attributes: MutableMap<String, Any> = HashMap()
        attributes["key 1"] = "value 1"
        val list: MutableList<String> = LinkedList()
        list.add("value 2")
        list.add("value 3")
        attributes["key 2"] = list
        Mockito.`when`(
            MParticle.getInstance()!!.Identity().currentUser!!.userAttributes
        ).thenReturn(attributes)
        val filteredAttributes = integration.allUserAttributes
        Assert.assertEquals("value 1", filteredAttributes["key 1"])
        Assert.assertEquals(list, filteredAttributes["key 2"])
    }

    @Test
    @Throws(Exception::class)
    fun testGetAllUserAttributesWithListsAndFilters() {
        MParticle.setInstance(MockMParticle())
        val integration: KitIntegration = AttributeListenerIntegration()
        val configuration = Mockito.mock(
            KitConfiguration::class.java
        )
        val mockArray = MockSparseBooleanArray()
        mockArray.put(MPUtility.mpHash("key 4"), false)
        mockArray.put(MPUtility.mpHash("key 3"), false)
        Mockito.`when`(configuration.userAttributeFilters).thenReturn(mockArray)
        integration.configuration = configuration
        val attributes: MutableMap<String, Any> = HashMap()
        attributes["key 1"] = "value 1"
        attributes["key 4"] = "value 4"
        val list: MutableList<String> = LinkedList()
        list.add("value 2")
        list.add("value 3")
        attributes["key 2"] = list
        attributes["key 3"] = list
        Mockito.`when`(
            MParticle.getInstance()!!.Identity().currentUser!!.userAttributes
        ).thenReturn(attributes)
        val filteredAttributes = integration.allUserAttributes
        Assert.assertEquals("value 1", filteredAttributes["key 1"])
        Assert.assertEquals(list, filteredAttributes["key 2"])
        Assert.assertNull(filteredAttributes["key 3"])
        Assert.assertNull(filteredAttributes["key 4"])
    }

    private inner class AttributeListenerIntegration : KitIntegration(), AttributeListener {
        override fun setUserAttribute(attributeKey: String, attributeValue: String) {}
        override fun setUserAttributeList(attributeKey: String, attributeValueList: List<String>) {}
        override fun supportsAttributeLists(): Boolean {
            return true
        }

        override fun setAllUserAttributes(
            userAttributes: Map<String, String>,
            userAttributeLists: Map<String, List<String>>
        ) {
        }

        override fun removeUserAttribute(key: String) {}
        override fun setUserIdentity(identityType: IdentityType, identity: String) {}
        override fun removeUserIdentity(identityType: IdentityType) {}
        override fun logout(): List<ReportingMessage>? {
            return null
        }

        override fun getName(): String? {
            return null
        }

        override fun onKitCreate(
            settings: Map<String, String>,
            context: Context
        ): List<ReportingMessage>? {
            return null
        }

        override fun setOptOut(optedOut: Boolean): List<ReportingMessage>? {
            return null
        }
    }
}
